src/s3/commands/deleteBucket.ts

```diff
@@ -0,0 +1,80 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+import { getLogger } from '../../shared/logger'
+import * as telemetry from '../../shared/telemetry/telemetry'
+import { localize } from '../../shared/utilities/vsCodeUtils'
+import { Commands } from '../../shared/vscode/commands'
+import { Window } from '../../shared/vscode/window'
+import { S3BucketNode } from '../explorer/s3BucketNode'
+import { S3Node } from '../explorer/s3Nodes'
+import { showErrorWithLogs } from '../util/messages'
+
+/**
+ * Deletes the bucket represented by the given node.
+ *
+ * Prompts the user for confirmation (user must type the bucket name).
+ * Deletes the bucket, showing a progress bar.
+ * Refreshes the parent node.
+ */
+export async function deleteBucketCommand(
+    node: S3BucketNode,
+    window = Window.vscode(),
+    commands = Commands.vscode()
+): Promise<void> {
+    getLogger().debug('DeleteBucket called for %O', node)
+
+    const isConfirmed = await showConfirmationDialog(node.bucket.name, window)
+    if (!isConfirmed) {
+        getLogger().info('DeleteBucket cancelled')
+        telemetry.recordS3DeleteBucket({ result: 'Cancelled' })
+        return
+    }
+
+    getLogger().info(`Deleting bucket ${node.bucket.name}`)
+    try {
+        await deleteWithProgress(node, window)
+
+        getLogger().info(`Successfully deleted bucket ${node.bucket.name}`)
+        telemetry.recordS3DeleteBucket({ result: 'Succeeded' })
+    } catch (e) {
+        getLogger().error(`Failed to delete bucket ${node.bucket.name}: %O`, e)
+        showErrorWithLogs(
+            localize('AWS.s3.deleteBucket.error.general', 'Failed to delete bucket {0}', node.bucket.name),
+            window
+        )
+        telemetry.recordS3DeleteBucket({ result: 'Failed' })
+    }
+
+    await refreshNode(node.parent, commands)
+}
+
+async function showConfirmationDialog(bucketName: string, window: Window): Promise<boolean> {
+    const prompt = localize('AWS.s3.deleteBucket.prompt', 'Enter {0} to confirm deletion', bucketName)
+    const confirmationInput = await window.showInputBox({
+        prompt,
+        placeHolder: bucketName,
+        validateInput: input => (input !== bucketName ? prompt : undefined),
+    })
+
+    return confirmationInput === bucketName
+}
+
+async function deleteWithProgress(node: S3BucketNode, window: Window): Promise<void> {
```

Comment:
Does this take enough time to require a progress bar/does the delete API even report deletion progress?

Reply:
Yes, it can definitely take a while. It has to iterate over all versions of all objects in the bucket (up to 1000 at a time) and then issue delete requests for all versions of all objects (again up to 1000 at a time).

It doesn't report explicit deletion progress, but this will show one of those infinite progress bars to show the user it's still in progress. You can see this briefly in the gif

Reply:
Oh gotcha, didn't realize that that would have to trigger individual deletions for each item in the bucket. Makes sense :+1: